internal/olm/operator/registry/index_image.go

```diff
@@ -124,14 +151,287 @@ func (c IndexImageCatalogCreator) CreateCatalog(ctx context.Context, name string
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	return content, nil
+}
+
+// setupFBCupdates starts the process of upgrading a bundle in an FBC. This function will recreate the FBC that was generated
+// during run bundle and upgrade a specific bundle in the specified channel.
+func runFBCupgrade(ctx context.Context, c *IndexImageCatalogCreator) error {
+	var err error
+
+	// render the index image if it is not the default index image
+	render := action.Render{Refs: []string{}}
+	if c.IndexImage != DefaultIndexImage {
+		render.Refs = append(render.Refs, c.IndexImage)
+	}
+
+	log.SetOutput(ioutil.Discard)
+	originalDeclcfg, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		return fmt.Errorf("error rendering index image %q: %v", c.IndexImage, err)
+	}
+
+	// add the upgraded bundle to the list of previous bundles so that they can be rendered with a single API call
+	c.PreviousBundles = append(c.PreviousBundles, c.BundleImage)
+	f := &FBCContext{
+		Package:     c.PackageName,
+		Refs:        c.PreviousBundles,
+		ChannelName: c.ChannelName,
+	}
+
+	// Adding the FBC "f" to the originalDeclcfg to generate a new FBC
+	declcfg, err := upgradeFBC(ctx, f, originalDeclcfg)
+	if err != nil {
+		return fmt.Errorf("error creating the upgraded FBC: %v", err)
+	}
+
+	// validate the declarative config and convert it to a string
+	var content string
+	if content, err = ValidateAndStringify(declcfg); err != nil {
+		return fmt.Errorf("error validating/stringifying the declarative config object: %v", err)
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	c.FBCContent = content
+
+	return nil
+}
+
+// upgradeFBC constructs a new File-Based Catalog from both the FBCContext object and the declarative config object. This function will check to see
+// if the FBCContext object "f" is already present in the original declarative config.
+func upgradeFBC(ctx context.Context, f *FBCContext, originalDeclCfg *declarativeconfig.DeclarativeConfig) (*declarativeconfig.DeclarativeConfig, error) {
+	var err error
+
+	// Rendering the bundle image and index image into declarative config format
+	render := action.Render{
+		Refs: f.Refs,
+	}
+
+	log.SetOutput(ioutil.Discard)
+	declcfg, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		return nil, fmt.Errorf("error in rendering the bundle and index image: %v", err)
+	}
+
+	// Ensuring a valid bundle size
+	if len(declcfg.Bundles) < 1 {
+		return nil, fmt.Errorf("bundle image should contain exactly one bundle blob")
+	}
+
+	// Checking if the existing file-based catalog (before upgrade) contains the bundle and channel that we intend to insert.
+	// If the bundle already exists, we error out. If the channel already exists, we store the index of the channel. This
+	// index will be used to access the channel from the declarative config object
+	channelExists := false
+	channelIndex := -1
```

Comment:
Can you remind me again why we are using the channel index? What purpose does it solve here?

Reply:
If the channel already exists, we are storing the index at which it exists in the Channels[] array of the declarative config object. Later, while creating entries for upgrading, we need to know which channel to insert these entries into (if the channel exists), which is why we're storing the index here.